src/extensions/default/PrefsCodeHints/main.js

```diff
@@ -0,0 +1,386 @@
+                };
+
+                // Traverse through child keys on of object and only add hintable keys.
+                if (preference.keys && preference.type === "object") {
+                    data[pref].keys = {};
+                    Object.keys(preference.keys).forEach(function (childPref) {
+                        if (preference.keys[childPref].hintable) {
+                            data[pref].keys[childPref] = preference.keys[childPref];
+                        }
+                    });
+                }
+            }
+        });
+    }
+    
+    /**
+     * Determines whether or not hints are available in the current context
+     * 
+     * @param {!Editor} editor
+     * @param {String} implicitChar
+     * @return {Boolean}
+     */
+    PrefsCodeHints.prototype.hasHints = function (editor, implicitChar) {
+        if (isPrefHintsEnabled && editor.getModeForSelection() === "application/json") {
+            this.editor = editor;
+            this.ctxInfo = JSONUtils.getContextInfo(this.editor, this.editor.getCursorPos(), true);
+            
+            if (this.ctxInfo && this.ctxInfo.tokenType) {
+                // Disallow hints for blacklisted keys.
+                if (this.ctxInfo.tokenType === JSONUtils.TOKEN_KEY &&
+                        parentKeyBlacklist.indexOf(this.ctxInfo.parentKeyName) !== -1) {
+                    return false;
+                }
+                return true;
+            }
+        }
+        return false;
+    };
+    
+    /**
+     * Returns a list of hints available in the current context
+     * 
+     * @param {String} implicitChar
+     * @return {!{hints: Array.<jQueryObject>, match: string, selectInitial: boolean, handleWideResults: boolean}}
+     */
+    PrefsCodeHints.prototype.getHints = function (implicitChar) {
+        var hints = [], ctxInfo, query, keys, values, option = {type: null, description: null, values: null};
+        
+        ctxInfo = this.ctxInfo = JSONUtils.getContextInfo(this.editor, this.editor.getCursorPos(), true);
+        
+        if (ctxInfo && ctxInfo.token) {
+            query = JSONUtils.stripQuotes(ctxInfo.token.string.substr(0, ctxInfo.offset)).trim();
+            if (JSONUtils.regexAllowedChars.test(query)) {
+                query = "";
+            }
+            
+            if (ctxInfo.tokenType === JSONUtils.TOKEN_KEY) {
+                // Provide hints for keys
+                
+                // Get options for parent key else use general options.
+                if (data[ctxInfo.parentKeyName] && data[ctxInfo.parentKeyName].keys) {
+                    keys = data[ctxInfo.parentKeyName].keys;
+                } else if (ctxInfo.parentKeyName === "language") {
+                    keys = LanguageManager.getLanguages();
+                    option.type = "object";
+                } else {
+                    keys = data;
+                }
+            
+                hints = $.map(Object.keys(keys), function (key) {
```

Comment:
`Array.prototype.map` will return undefined if it doesn't match. A possible workaround would be to use a `Array.prototype.filter` on the array returned by `map` to exclude `undefined`. $.map works fine for this scenario.